### PR TITLE
CDAP-8445 return 503 for enqueue endpoint is dataset unavailable

### PR DESCRIPTION
**Pull request by albertshau**
Fixes the stream service endpoint for enqueuing messages to a
stream so that it returns a 503 if some dependent service
(dataset service in this case) is not available.

Also changing the DefaultStreamWriter so that it retries
if it gets a 503 back from the stream service.

### CODE REVIEW

**Comment by albertshau** [comment]
http://builds.cask.co/browse/CDAP-RUT585

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamHandler.java [hunk 1]
```diff
@@ -391,34 +374,30 @@ public void emitMetrics(long bytesWritten, long eventsWritten) {
...
     Reader reader = new InputStreamReader(new ChannelBufferInputStream(request.getContent()));
     StreamProperties properties;
     try {
       properties = GSON.fromJson(reader, StreamProperties.class);
     } catch (Exception e) {
-      responder.sendString(HttpResponseStatus.BAD_REQUEST, "Invalid stream configuration. Please check that the " +
-        "configuration is a valid JSON Object with a valid schema.");
-      return null;
+      throw new BadRequestException("Invalid stream configuration. Please check that the " +
+                                      "configuration is a valid JSON Object with a valid schema.");
```
Comment (sagarkapare) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamHandler.java]: Should we add information from `e` to the `BadRequestException`?

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamHandler.java [hunk 2]
```diff
@@ -391,34 +374,30 @@ public void emitMetrics(long bytesWritten, long eventsWritten) {
...
     }
 
     // Validate ttl
     Long ttl = properties.getTTL();
     if (ttl != null && ttl < 0) {
-      responder.sendString(HttpResponseStatus.BAD_REQUEST, "TTL value should be positive.");
-      return null;
+      throw new BadRequestException("TTL value should be positive.");
```
Comment (sagarkapare) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamHandler.java]: better add `ttl` in the message.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamHandler.java [hunk 3]
```diff
@@ -429,21 +408,16 @@ private StreamProperties getAndValidateConfig(HttpRequest request, HttpResponder
...
+        throw new BadRequestException("Invalid format, unable to instantiate format " + formatName);
       }
     }
 
     // Validate notification threshold
     Integer threshold = properties.getNotificationThresholdMB();
     if (threshold != null && threshold <= 0) {
-      responder.sendString(HttpResponseStatus.BAD_REQUEST, "Threshold value should be greater than zero.");
-      return null;
+      throw new BadRequestException("Threshold value should be greater than zero.");
```
Comment (sagarkapare) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/service/StreamHandler.java]: Better add `threshold` in the message.

**Review by sagarkapare — APPROVED** [review]
Minor comments. otherwise LGTM 👍

**Comment by albertshau** [comment]
only changed error messages. merging without another build.

**Pull request merged**